### PR TITLE
Use JUnit 5 for the JVM Tests

### LINKED ISSUE #2273: Test result reporting broken with Surefire 3.0-M4 and later
### Describe the bug
When running a student exercise with Maven Surefire, we ran into a regression. Starting from a reporting refactoring in Surefire at v3.0-M4, the message of the failure does not get reported any more in the XML output. We will file an issue on their side for this, as we think this is a bug in Surefire. However, this has consequences breaking Artemis. The non-existing message in the XML leads to a null message in the Json of the Artemis Jenkins plugin (v1.0). This again is picked up in Artemis as successful test run - despite the run being reported correctly as failure.

#### To Reproduce
1. Create a student exercise with Surefire >= 3.0-M4
2. Add failing tests to it
3. Run them via Artemis (either in a Student exercise instance or in the template repo)

#### Expected behavior
Even if no message gets passed, failed or errored tests should not be recorded as "success". I expect them to be recorded as "failed" or "error". The missing message might be replaced with a default warning message, stating that no message was received and an instructor should check the setup.

### Environment
<details><pre>
 - OS: macOS client, Ubuntu 20.04LTS server
 - Browser: Firefox
 - Version: latest dev
</pre></details>

#### Additional context
Java student exercise using Maven Surefire >= 3.0-M4 as test execution and reporting engine. Using JUnit-Platform (i.e. Junit 5 runner) as test engine

### PR DESCRIPTION
**Pull request by MaisiKoleni**
### Checklist
- [x] I tested *all* changes and *all* related features with different users (student, tutor, instructor, admin) on the test server https://artemistest.ase.in.tum.de.
- [x] Server: I documented the Java code using JavaDoc style.

### Motivation and Context
Switch to JUnit 5 and newer Maven Surefire version to support https://github.com/ls1intum/artemis-java-test-sandbox and also resolves #2273.

The name of the test cases in Artemis will **not** change.

### Description
- Use `artemis-java-test-sandbox` already as provider for JUnit 5, AssertJ and more. Remove JUnit 4.
- Use Maven Surefire version `3.0.0-M5`
- Adjust the Java test template to JUnit 5 Jupiter using dynamic tests.

#### TODO:
- [x] Change version and check some JavaDoc
- [x] Also update Kotlin (and #2278 is also affected)

### Steps for Testing
0. Review the code (note that the Kotlin code is the same apart from `FunctionalTest`)
1. Create a Java (and also Kotlin) programming exercise as instructor (with or without SCA should not matter) on TS1/TS3
2. Check that both solution and template have **13** test cases, with the solution passing all of them and the template none
3. Participate in some way and check that the tests still work. Check passing at least _one behavior test_ (e.g. sorting functions) and _one structural_ (e.g. adding the sort interface).
4. Check that the tasks **and** the class diagram in the problem statement are updated correctly
5. _Repeat that on TS3 which uses another build system (there might still be problems due to #2273 and #2224 regarding failing test)_

### Test Coverage

_Unchanged._

### CODE REVIEW

File: src/main/resources/templates/kotlin/test/testutils/ClassNameScanner.java [hunk 1]
```diff
@@ -219,25 +224,25 @@ private ScanResult computeScanResult() {
...
      * @param assignmentFolderName: The root folder where the method starts walking the project structure.
      * @param node: The current node the method is visiting.
-     * @param foundTypes: The JSON object where the type names and packages get appended.
+     * @param types: The JSON object where the type names and packages get appended.
      */
     private void walkProjectFileStructure(String assignmentFolderName, File node, Map<String, List<String>> foundTypes) {
         String fileName = node.getName();
 
-        if(fileName.contains(".java") || fileName.contains(".kt")) {
+        if(fileName.contains(".java")) {
```
Comment (krusche) [inline src/main/resources/templates/kotlin/test/testutils/ClassNameScanner.java:232]: re-add kt

File: src/main/resources/templates/kotlin/test/testutils/ClassNameScanner.java [hunk 2]
```diff
@@ -219,25 +224,25 @@ private ScanResult computeScanResult() {
...
             String[] fileNameComponents = fileName.split("\\.");
 
             String className = fileNameComponents[fileNameComponents.length - 2];
-            int until = node.getPath().indexOf(".java");
-            if (until == -1) {
-                until = node.getPath().indexOf(".kt");
-            }
-            String packageName = node.getPath().substring(0, until);
+            String packageName = node.getPath().substring(0, node.getPath().indexOf(".java"));
```
Comment (krusche) [inline src/main/resources/templates/kotlin/test/testutils/ClassNameScanner.java:236]: re-add kt

File: src/main/resources/templates/kotlin/test/testFiles/structural/ConstructorTest.java [hunk 3]
```diff
@@ -1,42 +1,40 @@
...
      * This method collects the classes in the structure oracle file for which constructors are specified.
-     * These classes are packed into a list, which represents the test data.
-     * @return A list of arrays containing each class' name, package and the respective JSON object defined in the structure oracle.
+     * These classes are then transformed into JUnit 5 dynamic tests.
+     * @return A dynamic test container containing the test for each class which is then executed by JUnit.
      */
-    @Parameterized.Parameters(name = "{0}")
-    public static Collection<Object[]> findClasses() {
-        List<Object[]> testData = new ArrayList<Object[]>();
+    @TestFactory
```
Comment (krusche) [inline src/main/resources/templates/kotlin/test/testFiles/structural/ConstructorTest.java:35]: can we add some timeout?

**Review by fde312 — CHANGES_REQUESTED** [review]
Tested it in ts2 and somehow all tests are successful even though I only solved bubble sort:
image

**Review by schultek — COMMENTED** [review]
This works for the bamboo setup on test server 1, but i had the same errors as @fde312 for the jenkins setup on ts2.

**Review by sleiss — APPROVED** [review]
LGTM

**Review by schultek — APPROVED** [review]
This works fine now.

**Review by kloessst — APPROVED** [review]
Tested multiple exercises with various configs. Works fine. Well done!

**Review by krusche — APPROVED** [review]
Thanks! Code looks good! Great work 👍🏻🚀

**Pull request merged**